plugins/circleci/src/circleci-config.ts

```diff
@@ -32,7 +33,8 @@ export default abstract class CircleCiConfigHook extends Hook<CircleCIState> {
   circleConfigPath = path.resolve(process.cwd(), '.circleci/config.yml')
   _circleConfig?: YAML.Document
   abstract job: string
-  jobOptions: JobConfig = {}
+  abstract jobOptions: JobConfig
+  additionalFields?: Record<string, unknown>
```

Comment:
So it’s a breaking change because of the assumptions we make about state initialisation. Initially, I tried to have a static method like `getInitialState` that would return the skeleton of the state – an object of empty objects or arrays typically. Unfortunately, TypeScript doesn’t support static abstract methods in classes at the moment, so we just initialise the state the first time the method is called instead.

This means that whatever subclass that's called first is the one that ends up being in charge of initialisation, and all the `install` implementations make assumptions about the structure of the initialised state. Now that we’ve added a new object `additionalFields` to the state, we have to set the empty object in the initialisation as all subsequent `install` code assumes it’s present. We can run into `TypeErrors`'s about it being `undefined` if it isn’t. Which subclass initialises the state is therefore important and adding a new field to the initial state will break the code as if older subclasses initialise the state first the new invariants won’t be upheld.

The state PR hasn’t been merged yet so this could be improved somehow, which might be worthwhile. Maybe we should avoid doing any state initialisation and always assume any object could be undefined? Or maybe there’s a more elegant alternative…?

Reply:
Thinking about this some more I don't think this has to be such a big deal. It should be okay to include an extra, one like
```ts
state.additionalFields ??= {}
```
to initialise the extra state field in case it wasn't included by another plugin. This is reasonable to make as a concession to avoid introducing a breaking change, and a comment can be left alongside it to note that it can be removed in the next major version. It's a little academic in this case anyway as I'm going to cherry pick the commit and put it in #299.